src/api/array_join.cc

```diff
@@ -0,0 +1,139 @@
+/**
+* Unless explicitly stated otherwise all files in this repository are licensed under the Apache-2.0 License.
+* This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+**/
+#include <new>
+#include <vector>
+#include <memory>
+#include <string>
+
+#include "array_join.h"
+#include "../tainted/range.h"
+#include "../tainted/string_resource.h"
+#include "../tainted/transaction.h"
+#include "../iast.h"
+
+using v8::FunctionCallbackInfo;
+using v8::Value;
+using v8::Local;
+using v8::Isolate;
+using v8::Object;
+using v8::String;
+
+using iast::tainted::Range;
+
+namespace iast {
+namespace api {
+
+const int DEFAULT_JOIN_SEPARATOR_LENGTH = 1;
+
+void copyRangesWithOffset(Transaction* transaction,
+        SharedRanges* origRanges,
+        SharedRanges* destRanges,
+        int offset) {
+    if (origRanges != nullptr) {
+        auto end = origRanges->end();
+        for (auto it = origRanges->begin(); it != end; it++) {
+            auto origRange = *it;
+            auto newRange = transaction->GetRange(
+                origRange->start + offset,
+                origRange->end + offset,
+                origRange->inputInfo,
+                origRange->secureMarks);
+            if (newRange != nullptr) {
+                destRanges->PushBack(newRange);
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+SharedRanges* getJoinResultRanges(Isolate* isolate,
+        Transaction* transaction, v8::Array* arr,
+        SharedRanges* separatorRanges,
+        int separatorLength) {
+    auto length = arr->Length();
+    int offset = 0;
+    auto newRanges = transaction->GetSharedVectorRange();
```

Comment:
We are getting and potentially creating a new SharedVector each time that `join` method is executed, even if it doesn't have any tainted object. Can we delay the initialization of `newRanges` to the point that we know that it will be necessary?